### ISSUE TITLE
ReadPreference is lost when connection is created

### ISSUE THREAD

**Issue by damienalexandre** [alcaeus/mongo-php-adapter#112]
There is an issue with the adapter when passing a custom readPreference.

The value is lost here: https://github.com/alcaeus/mongo-php-adapter/blob/f37daacac4aaf3feae2312d93dc79ce8d399684e/lib/Mongo/MongoDB.php#L69
Because `getReadPreference` return `null`.

Using the new MondoDB extension directly work as expected. And it worked with the old extension too.

### How to reproduce

You have to setup a replicaSet of course.

```php
$url = "mongodb://127.0.0.1:27017,127.0.0.1:27018";
$params = ['username' => 'test', 'password' => 'test', 'db' => 'test', 'readPreference' => 'secondaryPreferred'];

$client = new \MongoClient($url, $params);

// Run a query to init everything
$cursor = $client->selectDB('stuffs')->selectCollection('stuffs')->find();
$cursor->next();

// It return mode 6, which is good
var_dump($client->getClient()->__debugInfo()['manager']->getReadPreference());
// It return the primary host, not good!
var_dump($cursor->info()['port']);
```

### How to quick-fix

I added this line right after the client creation to tell which readPreference I wanted:

```php
$client = new \MongoClient($url, $params);
$client->setReadPreference('secondaryPreferred');
```

### Why not doing a PR?

Because I'm not sure of the best way to fix this behavior, this option can be set in two places:
- in the server url: `127.0.0.1:27017?readPreference=secondaryPreferred`
- in the client option like in my exemple
And I didn't found any code doing parameters lookup in this adapter.

I can work on it if a direction is given.

thanks! :beers:

**Comment by alcaeus** [comment]
Thanks for the report. You are correct, the options may be set using the connection URI as well as the options argument, so `MongoClient::__construct` would have to parse these and apply readPreference and writeConcern (and potentially other options) as well to make sure the return values are correct.

**Comment by damienalexandre** [comment]
Thanks for your answer, I'm not available this weekend so you can take the lead on it.

Some thoughts:

- both the url and the params needs to be checked;
- I didn't find the source but I think the params from the option array overwrite the ones from the url if present in both;
- only some options were supported by the old client, see http://php.net/manual/en/mongoclient.construct.php for the full list (and maybe have the default values set too).

Thanks!

**Comment by alcaeus** [comment]
Just a quick update since I actually planned to have this done over the weekend: I realized that parsing the query string of the connection URL is a bit more complicated since users may specify multiple values for `readPreferenceTags`. This is described in example 1 in the [read preference documentation](https://secure.php.net/manual/en/mongo.readpreferences.php):
```
// Prefer the nearest server in the "east" data center, then a server in the
// "west" data center, and finally fall back to no tag set preference
$uri  = 'mongodb://rs1.example.com,rs2.example.com/';
$uri .= '?readPreference=nearest';
$uri .= '&readPreferenceTags=dc:east';
$uri .= '&readPreferenceTags=dc:west';
$uri .= '&readPreferenceTags=';
```

I discovered this when testing so I'll have to rewrite the functionality and manually parse the query string instead of relying on `parse_str` to do it. If this bug is posing a problem for you, I suggest not passing the readPreference via options or the URL string and instead call `setReadPreference` on the `MongoClient` instance after creating it.
